[site-isolation] Send pageScaleFactor changes to all relevant processes.

mattwoodrow opened the pull request:
#### 1e0f90c0a4748d952997071013e3836e5306dd1d
<pre>
[site-isolation] Send pageScaleFactor changes to all relevant processes.
<a href="https://bugs.webkit.org/show_bug.cgi?id=280040">https://bugs.webkit.org/show_bug.cgi?id=280040</a>
&lt;<a href="https://rdar.apple.com/136338665">rdar://136338665</a>&gt;

Reviewed by Charlie Wolfe.

Page scale changes can be initiated from both the WebProcess, as well as the UI process.
Adds mechanisms to share to all processes from the UI process, as well as sharing to all other web process when initiated from there.

Changes the TestRunner setter to use the same WebPageProxy setter as other API endpoints do, and adds a new test using a site-isolated subframe.

Also fixes Page::setPageScaleFactor to handle having multiple (and non-main) root frames.

* LayoutTests/http/tests/site-isolation/page-scale-expected.html: Added.
* LayoutTests/http/tests/site-isolation/page-scale.html: Added.
* LayoutTests/http/tests/site-isolation/resources/green-square.html: Added.
* Source/WebCore/loader/HistoryController.cpp:
(WebCore::HistoryController::restoreScrollPositionAndViewState):
* Source/WebCore/page/Page.cpp:
(WebCore::Page::setPageScaleFactor):
* Source/WebCore/rendering/RenderLayerBacking.cpp:
(WebCore::RenderLayerBacking::RenderLayerBacking):
(WebCore::RenderLayerBacking::createPrimaryGraphicsLayer):
* Source/WebCore/rendering/RenderLayerBacking.h:
* Source/WebCore/rendering/RenderLayerCompositor.cpp:
(WebCore::RenderLayerCompositor::ensureRootLayer):
* Source/WebCore/rendering/RenderObject.cpp:
(WebCore::printPaintOrderTreeForLiveDocuments):
(WebCore::printRenderTreeForLiveDocuments):
(WebCore::printLayerTreeForLiveDocuments):
(WebCore::printGraphicsLayerTreeForLiveDocuments):
* Source/WebKit/UIProcess/API/C/WKPage.cpp:
(WKPageSetScaleFactor):
(WKPageSetPageScaleFactorForTesting):
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewTesting.mm:
(-[WKWebView _setPageScale:withOrigin:]):
* Source/WebKit/UIProcess/ViewGestureController.cpp:
(WebKit::ViewGestureController::applyMagnification):
(WebKit::ViewGestureController::endMagnificationGesture):
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::scalePage):
(WebKit::WebPageProxy::scalePageInViewCoordinates):
(WebKit::WebPageProxy::scalePageRelativeToScrollPosition):
(WebKit::WebPageProxy::scaleView):
(WebKit::WebPageProxy::pageScaleFactorDidChange):
(WebKit::WebPageProxy::viewScaleFactorDidChange):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/WebPageProxy.messages.in:
* Source/WebKit/UIProcess/WebPageProxyTesting.cpp:
(WebKit::WebPageProxyTesting::setPageScaleFactor): Deleted.
* Source/WebKit/UIProcess/WebPageProxyTesting.h:
* Source/WebKit/UIProcess/mac/WKFullScreenWindowController.mm:
(-[WKFullScreenWindowController enterFullScreen:]):
(-[WKFullScreenWindowController finishedEnterFullScreenAnimation:]):
(-[WKFullScreenWindowController finishedExitFullScreenAnimationAndExitImmediately:]):
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::didScaleView):
(WebKit::WebPage::scaleView):
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/WebPage.messages.in:
* Source/WebKit/WebProcess/WebPage/WebPageTesting.cpp:
(WebKit::WebPageTesting::setPageScaleFactor): Deleted.
* Source/WebKit/WebProcess/WebPage/WebPageTesting.h:
* Source/WebKit/WebProcess/WebPage/WebPageTesting.messages.in:

Canonical link: <a href="https://commits.webkit.org/284435@main">https://commits.webkit.org/284435@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c3b67a684ef1be4ccf99cfa29e5d20c0c91ef6d0

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/69282 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/48682 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/21955 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/73364 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/20440 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/71399 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/56483 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/20289 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/55101 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/13569 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/72348 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/44420 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/59812 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/35580 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/41087 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/17242 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/18815 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/63028 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/17587 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/75076 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/13264 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/16817 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/62766 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/13303 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/59895 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/62672 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/15393 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/10690 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/4299 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/44486 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/45560 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/46755 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/45301 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

achristensen07 commented on Source/WebKit/WebProcess/WebPage/WebPage.cpp:
```diff
@@ -7650,7 +7677,7 @@ void WebPage::didCommitLoad(WebFrame* frame)
 
     // Only restore the scale factor for standard frame loads (of the main frame).
-    if (frame->coreLocalFrame()->loader().loadType() == FrameLoadType::Standard) {
+    if (frame->isMainFrame() && frame->coreLocalFrame()->loader().loadType() == FrameLoadType::Standard) {
```
This looks like a change for a separate PR.

achristensen07 commented on Source/WebCore/testing/Internals.cpp:
```diff
@@ -571,6 +570,7 @@ void Internals::resetToConsistentState(Page& page)
         return;
 
     localMainFrame->setTextZoomFactor(1.0f);
+    page.chrome().client().setPageScaleFactor(1, IntPoint(0, 0), [] { });
```
I think this should be removed and completely replaced by UI process initiated resetting.

achristensen07 commented on Source/WebCore/testing/Internals.h:
```diff
@@ -607,6 +607,7 @@ class Internals final
     ExceptionOr<String> pageSizeAndMarginsInPixels(int pageNumber, int width, int height, int marginTop, int marginRight, int marginBottom, int marginLeft) const;
 
     ExceptionOr<float> pageScaleFactor() const;
+    void setPageScaleFactor(float scaleFactor, DOMPromiseDeferred<void>&&);
```
This was intentionally moved to testRunner and shouldn't be re-added here.

mattwoodrow replied:
It was, but I think there's value in having it.

We have quite a few places that originate a page scale change within the WebContent process, and then we take a slightly different code path to share it to all the other WebContent process.

From what I can see: WKBundlePageSetScaleAtOrigin, HistoryController, DrawingArea::scaleViewToFitDocumentIfNeeded and RemoteLayerTreeDrawingAreaMac::applyTransientZoomToPage are all examples of that.

As long as we have those, it seems valuable to also have a matching entry point for test code.

Maybe there's a way to get rid of all of those, and then we can avoid having the impl code as well as the test code?

`kWKLayoutModeDynamicSizeComputedFromMinimumDocumentSize` appears to be API, but maybe it's not used.

achristensen07 commented on Source/WebCore/loader/HistoryController.cpp:
```diff
@@ -161,7 +163,7 @@ void HistoryController::restoreScrollPositionAndViewState()
         LOG(Scrolling, "HistoryController::restoreScrollPositionAndViewState scrolling to %d,%d", desiredScrollPosition.x(), desiredScrollPosition.y());
         if (page && frame->isMainFrame() && currentItem->pageScaleFactor())
-            page->setPageScaleFactor(currentItem->pageScaleFactor() * page->viewScaleFactor(), desiredScrollPosition);
+            page->chrome().client().setPageScaleFactor(currentItem->pageScaleFactor(), desiredScrollPosition, [] { });
```
I'm not sure about this.  It's pretty tangled up right now, but I think we want to move it to the UI process being in charge of setting this.  Probably in the future once the history in general is more UI process based.  Charlie's working on that.

charliewolfe commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -5248,7 +5251,9 @@ void WebPageProxy::scalePageInViewCoordinates(double scale, const IntPoint& cent
 
-    send(Messages::WebPage::DidScalePageInViewCoordinates(scale, centerInViewCoordinates));
+    forEachWebContentProcess([&] (auto& process, auto pageID) {
+        process.send(Messages::WebPage::DidScalePageInViewCoordinates(scale, centerInViewCoordinates), pageID);
```
This change doesn’t look like it does anything? The receiver still has a local mainframe downcast and early return.

mattwoodrow replied:
This is... a good point. Fixed that :)

charliewolfe commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -5260,7 +5265,9 @@ void WebPageProxy::scalePageRelativeToScrollPosition(double scale, const IntPoin
 
-    send(Messages::WebPage::DidScalePageRelativeToScrollPosition(scale, origin));
+    forEachWebContentProcess([&] (auto& process, auto pageID) {
+        process.send(Messages::WebPage::DidScalePageRelativeToScrollPosition(scale, origin), pageID);
```
Same thing for this one.

charliewolfe commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -5662,7 +5671,32 @@ static bool scaleFactorIsValid(double scaleFactor)
+    RefPtr process = dynamicDowncast<WebProcessProxy>(WebProcessProxy::fromConnection(connection));
+    MESSAGE_CHECK_BASE(process, connection);
+    if (process != &legacyMainFrameProcess())
+        return;
```
This could just be:
```suggestion
    if (!legacyMainFrameProcess().hasConnection(connection))
        return;
```

charliewolfe commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -5662,7 +5671,32 @@ static bool scaleFactorIsValid(double scaleFactor)
+    RefPtr process = dynamicDowncast<WebProcessProxy>(WebProcessProxy::fromConnection(connection));
+    MESSAGE_CHECK_BASE(process, connection);
+    if (process != &legacyMainFrameProcess())
+        return;
```
```suggestion
    if (!legacyMainFrameProcess().hasConnection(connection))
        return;
```

mattwoodrow committed:
[site-isolation] Send pageScaleFactor changes to all relevant processes.
https://bugs.webkit.org/show_bug.cgi?id=280040
<rdar://136338665>

Reviewed by Charlie Wolfe.

Page scale changes can be initiated from both the WebProcess, as well as the UI process.
Adds mechanisms to share to all processes from the UI process, as well as sharing to all other web process when initiated from there.

Changes the TestRunner setter to use the same WebPageProxy setter as other API endpoints do, and adds a new test using a site-isolated subframe.

Also fixes Page::setPageScaleFactor to handle having multiple (and non-main) root frames.

* LayoutTests/http/tests/site-isolation/page-scale-expected.html: Added.
* LayoutTests/http/tests/site-isolation/page-scale.html: Added.
* LayoutTests/http/tests/site-isolation/resources/green-square.html: Added.
* Source/WebCore/loader/HistoryController.cpp:
(WebCore::HistoryController::restoreScrollPositionAndViewState):
* Source/WebCore/page/Page.cpp:
(WebCore::Page::setPageScaleFactor):
* Source/WebCore/rendering/RenderLayerBacking.cpp:
(WebCore::RenderLayerBacking::RenderLayerBacking):
(WebCore::RenderLayerBacking::createPrimaryGraphicsLayer):
* Source/WebCore/rendering/RenderLayerBacking.h:
* Source/WebCore/rendering/RenderLayerCompositor.cpp:
(WebCore::RenderLayerCompositor::ensureRootLayer):
* Source/WebCore/rendering/RenderObject.cpp:
(WebCore::printPaintOrderTreeForLiveDocuments):
(WebCore::printRenderTreeForLiveDocuments):
(WebCore::printLayerTreeForLiveDocuments):
(WebCore::printGraphicsLayerTreeForLiveDocuments):
* Source/WebKit/UIProcess/API/C/WKPage.cpp:
(WKPageSetScaleFactor):
(WKPageSetPageScaleFactorForTesting):
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewTesting.mm:
(-[WKWebView _setPageScale:withOrigin:]):
* Source/WebKit/UIProcess/ViewGestureController.cpp:
(WebKit::ViewGestureController::applyMagnification):
(WebKit::ViewGestureController::endMagnificationGesture):
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::scalePage):
(WebKit::WebPageProxy::scalePageInViewCoordinates):
(WebKit::WebPageProxy::scalePageRelativeToScrollPosition):
(WebKit::WebPageProxy::scaleView):
(WebKit::WebPageProxy::pageScaleFactorDidChange):
(WebKit::WebPageProxy::viewScaleFactorDidChange):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/WebPageProxy.messages.in:
* Source/WebKit/UIProcess/WebPageProxyTesting.cpp:
(WebKit::WebPageProxyTesting::setPageScaleFactor): Deleted.
* Source/WebKit/UIProcess/WebPageProxyTesting.h:
* Source/WebKit/UIProcess/mac/WKFullScreenWindowController.mm:
(-[WKFullScreenWindowController enterFullScreen:]):
(-[WKFullScreenWindowController finishedEnterFullScreenAnimation:]):
(-[WKFullScreenWindowController finishedExitFullScreenAnimationAndExitImmediately:]):
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::didScaleView):
(WebKit::WebPage::scaleView):
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/WebPage.messages.in:
* Source/WebKit/WebProcess/WebPage/WebPageTesting.cpp:
(WebKit::WebPageTesting::setPageScaleFactor): Deleted.
* Source/WebKit/WebProcess/WebPage/WebPageTesting.h:
* Source/WebKit/WebProcess/WebPage/WebPageTesting.messages.in:

Canonical link: https://commits.webkit.org/284435@main
